p2panda-net/src/sync/manager.rs

```diff
@@ -3,45 +3,154 @@
+use tracing::{debug, error, warn};
 
 use crate::engine::ToEngineActor;
 use crate::sync::{self, SYNC_CONNECTION_ALPN};
 
+const MAX_RETRY_ATTEMPTS: u8 = 5;
+
+/// A newly discovered peer and topic combination to be sent to the sync manager.
 #[derive(Debug)]
-pub struct SyncManager {
-    known_peer_topics: HashMap<NodeId, HashSet<TopicId>>,
+pub struct ToSyncManager {
+    peer: NodeId,
+    topics: Vec<TopicId>,
+}
+
+impl ToSyncManager {
+    pub(crate) fn new(peer: NodeId, topics: Vec<TopicId>) -> Self {
+        Self { peer, topics }
+    }
+}
+
+#[derive(Debug)]
+struct SyncAttempt {
+    peer: NodeId,
+    topic: TopicId,
+    attempts: u8,
+}
+
+impl SyncAttempt {
+    fn new(peer: NodeId, topic: TopicId) -> Self {
+        Self {
+            peer,
+            topic,
+            attempts: 0,
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+enum SyncAttemptError {
+    #[error("connection error")]
+    Connection,
+    #[error("sync error")]
+    Sync,
+}
+
+/// An API for scheduling outbound connections and sync attempts.
+#[derive(Debug)]
+pub(crate) struct SyncManager {
     active_sync_sessions: HashMap<TopicId, HashSet<NodeId>>,
     completed_sync_sessions: HashMap<TopicId, HashSet<NodeId>>,
-    engine_actor_tx: Sender<ToEngineActor>,
     endpoint: Endpoint,
+    engine_actor_tx: Sender<ToEngineActor>,
+    inbox: Receiver<ToSyncManager>,
+    known_peer_topics: HashMap<NodeId, HashSet<TopicId>>,
+    max_retry_attempts: u8,
     sync_protocol: Arc<dyn for<'a> SyncProtocol<'a> + 'static>,
+    sync_queue_tx: Sender<SyncAttempt>,
+    sync_queue_rx: Receiver<SyncAttempt>,
 }
 
 impl SyncManager {
-    pub fn new(
+    /// Create a new instance of the `SyncManager` and return it along with a channel sender.
+    pub(crate) fn new(
         endpoint: Endpoint,
         engine_actor_tx: Sender<ToEngineActor>,
         sync_protocol: Arc<dyn for<'a> SyncProtocol<'a> + 'static>,
-    ) -> Self {
-        Self {
-            known_peer_topics: HashMap::new(),
+    ) -> (Self, Sender<ToSyncManager>) {
+        let (sync_queue_tx, sync_queue_rx) = mpsc::channel(128);
+        let (sync_manager_tx, sync_manager_rx) = mpsc::channel(256);
+
+        let sync_manager = Self {
             active_sync_sessions: HashMap::new(),
             completed_sync_sessions: HashMap::new(),
-            engine_actor_tx,
             endpoint,
+            engine_actor_tx,
+            inbox: sync_manager_rx,
+            known_peer_topics: HashMap::new(),
+            max_retry_attempts: MAX_RETRY_ATTEMPTS,
             sync_protocol,
+            sync_queue_tx,
+            sync_queue_rx,
+        };
+
+        (sync_manager, sync_manager_tx)
+    }
+
+    /// Add a peer and topic combination to the sync connection queue.
+    async fn schedule_attempt(&mut self, peer: NodeId, topic: TopicId) -> Result<()> {
+        let sync_attempt = SyncAttempt::new(peer, topic);
+        self.sync_queue_tx.send(sync_attempt).await?;
```

Comment:
> This will block when the queue is full; I don't think that's a problem for now but something to bear in mind

Is it possible to never make it too full? If we know we only can handle max. 10 sync sessions at a time then we should a) never push to the queue when it's full and b) the queue can not be larger than 10?

It could be otherwise a potential bug source which might be hard to find in the future (unexpected blocking behaviour).